staging/src/k8s.io/apiserver/pkg/authorization/authorizer/interfaces.go

```diff
@@ -144,3 +144,11 @@ func (a AttributesRecord) IsResourceRequest() bool {
 func (a AttributesRecord) GetPath() string {
 	return a.Path
 }
+
+type Decision int
+
+const (
+	DecisionDeny Decision = iota
```

Comment:
Seems like `DecisionDisallow` would be equivalent to `DecisionNoOpinion`. Is `DecisionUnknown` needed if errors are communicated out-of-band?

Reply:
> Seems like DecisionDisallow would be equivalent to DecisionNoOpinion

Agree. `DecisionDisallow` and `DecisionNoOpinion` are the same if we deny by default, which we do.

> DecisionUnknown  // I error'd so I cannot tell you what I think

I think I'd expect the decision to be returned along with any evaluation error (webhook could be configured to fail to deny, or fail to noopinion, etc)

Reply:
I know the meaning of the decision can be derived by the surrounding context (i.e. we deny by default or an error was returned), but why should we not just encode it all in the decision itself?  Especially if we include the decision in the SAR response.

Reply:
If we're going to differentiate between failing open vs failing closed, we probably want to let SAR be explicit between "disallow" and "internal error", right?

```
type Decision string

var (
	DecisionDeny Decision = "decisionDeny"  // short-circuit fail, no one is allowed to say yes to this
        DecisionDisallow = "decisionDisallow"  // I say no but its OK if someone else says yes
 	// I error'd so I cannot tell you what I think. If fail closed, this is equivalent to DecisionDeny, otherwise
	// it acts as DecisionDisallow.
	DecisionUnknown = "decisionUnknown" 
)

type SubjectAccessReviewStatus struct {
	// Allowed is required.  True if the action would be allowed, false otherwise.
	Allowed bool
	// Decision is optional.  If provided it describes the behavior of a denial letting authorizers unequivocally deny
	// the action even if another authorizer would allow it.
	Denial Decision

	Reason string
	EvaluationError string
}
```

Reply:
> we probably want to let SAR be explicit between "disallow" and "internal error", right?

I don't think so. A webhook authorizer has to determine what to do *if it does not get a SAR response back*, but it seems odd to obtain a SAR response without an actual decision, and have to combine that with a locally decided continue/fail policy. That would mean clients using the same external SAR server could react differently to the same SAR response, which I think weakens the authorization delegation.